rotkehlchen/exchanges/woo.py

```diff
@@ -0,0 +1,404 @@
+                    error=str(e),
+                )
+                self.msg_aggregator.add_error(
+                    'Failed to deserialize a Woo balance. '
+                    'Check logs for details. Ignoring it.',
+                )
+                continue
+            except (UnknownAsset, UnsupportedAsset) as e:
+                log.error(str(e))
+                asset_tag = 'unknown' if isinstance(e, UnknownAsset) else 'unsupported'
+                self.msg_aggregator.add_warning(
+                    f'Found {asset_tag} Woo asset {e.identifier}. Ignoring its balance query.',
+                )
+                continue
+            try:
+                usd_price = Inquirer().find_usd_price(asset=asset)
+            except RemoteError as e:
+                log.error(str(e))
+                self.msg_aggregator.add_error(
+                    f'Error processing Woo balance result due to inability to '
+                    f'query USD price: {e!s}. Skipping balance entry.',
+                )
+                continue
+
+            assets_balance[asset] = Balance(
+                amount=amount,
+                usd_value=amount * usd_price,
+            )
+
+        return assets_balance, ''
+
+    def query_online_trade_history(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> tuple[list[Trade], tuple[Timestamp, Timestamp]]:
+        """Return trade history on Woo in a range of time."""
+        options = {
+            'limit': API_MAX_LIMIT,
+        }
+
+        trades: list[Trade] = self._api_query_paginated(
+            start_ts=start_ts,
+            end_ts=end_ts,
+            options=options,
+            case='/v1/client/trades',
+        )
+        return trades, (start_ts, end_ts)
+
+    def validate_api_key(self) -> tuple[bool, str]:
+        """Validates that the Woo API key is good for usage in rotki"""
+        response = self._api_query('v1/client/trades')
+
+        if response.status_code != HTTPStatus.OK:
+            result, msg = self._process_unsuccessful_response(
+                response=response,
+                case='validate_api_key',
+            )
+            return result, msg
+
+        return True, ''
+
+    def _api_query(
+            self,
+            endpoint: str,
+            method: Literal['GET', 'POST'] = 'GET',
+            options: Optional[dict[str, Any]] = None,
+    ) -> Response:
+        """Request a  Woo API v2 endpoint (from `endpoint`).
+        """
+        call_options = options.copy() if options else {}
+        query_params = call_options if method == 'GET' else {}
+        data = call_options if method == 'POST' else None
+        request_url = f'{self.base_uri}/{endpoint}'
+        timestamp = str(ts_now_in_ms())
+
+        parameters = '&'.join([f'{key}={value}' for key, value in sorted(query_params.items())])
+        if data:
+            data = '&'.join([f'{key}={value}' for key, value in sorted(data.items())])
```

Comment:
you can use urllib to do this `&` thing

```
urllib.parse.urlencode(params)
```

not sure why you need the sorted thing. Also notice that dicts keep insertion order when read